File: Backend Codes/MATLAB final version/calcTouRate.m
function [Cbuy]= calcTouRate(year,onPrice, midPrice, offPrice, onHours, midHours, offHours, Month, Day, holidays)
          
startDate = datetime(year, 1, 1);
firstDayOfWeek = weekday(startDate); % Get the day of the week for the first day
Cbuy = zeros(1, 8760);

for m = 1:12
    
    t_start = 24 * sum(Day(1:m - 1)) + 1;
    t_end = 24 * sum(Day(1:m));
    t_index = t_start:t_end;
    nt = length(t_index);
    
    if Month(m) == 1 % for summer
        
        tp = onHours(1, :);
        tm = midHours{1};
        toff = offHours{1};
        P_peak = onPrice(1);
        P_mid = midPrice(1);
        P_offpeak = offPrice(1);
        
    else % for winter
        tp = onHours(2, :);
        tm = midHours{2};
        toff = offHours{2};
        P_peak = onPrice(2);
        P_mid = midPrice(2);
        P_offpeak = offPrice(2);
    end
    Cbuy(t_index) = P_offpeak; % set all hours to offpeak by default
    
    for d = 1:Day(m)
        idx0 = t_index(tp) + 24 * (d - 1);
        Cbuy(idx0) = P_peak;
        idx1 = t_index(tm) + 24 * (d - 1);
        Cbuy(idx1) = P_mid;
    end
end

for d = 1:365
    currentDate = startDate + days(d - 1);
    currentDayOfWeek = weekday(currentDate);
    
    if currentDayOfWeek == 1 || currentDayOfWeek == 7 % Saturday or Sunday
        st = 24 * (d - 1) + 1;
        ed = 24 * d;
        Cbuy(st:ed) = P_offpeak;
    end
    
    if ismember(d, holidays)
        st = 24 * (d - 1) + 1;
        ed = 24 * d;
        Cbuy(st:ed) = P_offpeak;
    end
end

File: Backend Codes/SAMA V1.03-GitHub/calcTouRate.py
from datetime import datetime, timedelta
import numpy as np

def calcTouRate(year, onPrice, midPrice, offPrice, onHours, midHours, season, daysInMonth, holidays):

  startDate = datetime(year, 1, 1)
  Cbuy = np.zeros(8760)
  tp = None
  tm= None

  for m in range(12):
    t_start = 24 * sum(daysInMonth[0:m])
    t_end = 24 * sum(daysInMonth[0:m + 1])
    t_index = np.arange(t_start, t_end)

    if season[m] == 1:  # for summer
      if len(onHours[season[m]-1]) > 0:
        tp = onHours[0]
      else:
        tp = None
      if len(midHours[season[m]-1]) > 0:
        tm = midHours[0]
      else:
        tm = None

      P_peak = onPrice[0]
      P_mid = midPrice[0]
      P_offpeak = offPrice[0]

    else:  # for winter
      if len(onHours[season[m]+1]) > 0:
        tp = onHours[1, :]
      else:
        tp = None
      if len(midHours[season[m]+1]) > 0:
        tm = midHours[1, :]
      else:
        tm = None

      P_peak = onPrice[1]
      P_mid = midPrice[1]
      P_offpeak = offPrice[1]

    Cbuy[t_index] = P_offpeak  # set all hours to off-peak by default

    for d in range(daysInMonth[m]):
      if tp is not None:
        for hour in tp:
          idx0 = t_index[hour-1] + 24 * d
          Cbuy[idx0] = P_peak

      if tm is not None:
        for hour in tm:
          idx1 = t_index[hour-1] + 24 * d
          Cbuy[idx1] = P_mid

    for d in range(365):
      currentDate = startDate + timedelta(days=d)
      currentMonth = currentDate.month -1  # months are 0 indexed in the season array
      if season[currentMonth] == 1:
        P_offpeak = offPrice[0]  # update off-peak price according to the current month's season
      else:
        P_offpeak = offPrice[1]

      currentDayOfWeek = currentDate.weekday()  # Monday is 0 and Sunday is 6

      if currentDayOfWeek == 5 or currentDayOfWeek == 6:  # Saturday or Sunday
        st = 24 * d
        ed = 24 * (d + 1)
        Cbuy[st:ed] = P_offpeak

      if d + 1 in holidays:
        st = 24 * d
        ed = 24 * (d + 1)
        Cbuy[st:ed] = P_offpeak

  return Cbuy